### PR TITLE
BREAKING CHANGE: update models to use `Set` rather than `List`

### LINKED ISSUE #159: [BREAKING] All optional lists become non-optional sets
@madpah a huge change i would suggest is: all optional lists become non-optional sets.
based on https://github.com/CycloneDX/cyclonedx-python-lib/pull/148/files#r798309225

this way a lot of problems would be solved: 
* no need for implementing `object.add_<list-property>()` all the time
* the order of the list-properties would no longer matter - no sorting, no whatsoever
* serializers/normalizers would always have a reliable non-optional data type to work with (remeber: an empty `set` is falsy - for easy size-checks)
* additionally: use the `set`-literal: instead of writing `set([...])` simply write `{...}` - if this was needed at any place, still

_Originally posted by @jkowalleck in https://github.com/CycloneDX/cyclonedx-python-lib/issues/148#issuecomment-1028715688_

### PR DESCRIPTION
**Pull request by madpah**
This PR looks to implement #159 for the reasons stated there.

As part of this, we have had to make our unit tests even better:
- We should not (but previously did) care about ordering of elements in either JSON or XML output for comparisons to be equal - ordering DOES NOT MATTER for equality
- Various updates made to adhere to [PEP-0008](https://www.python.org/dev/peps/pep-0008/) along the way

Resolves #159

### CODE REVIEW

File: cyclonedx/model/issue.py [hunk 1]
```diff
@@ -99,181 +99,99 @@ class IssueType:
...
-                name=source_name, url=source_url
-            )
+    def __init__(self, *, classification: IssueClassification, id_: Optional[str] = None, name: Optional[str] = None,
+                 description: Optional[str] = None, source: Optional[IssueTypeSource] = None,
+                 references: Optional[Iterable[XsUri]] = None) -> None:
+        self.type = classification
+        self.id = id_
+        self.name = name
+        self.description = description
+        self.source = source
```
[the remark this answers is not in the document]
Comment (madpah) [reply cyclonedx/model/issue.py:109]: Good catch @sonatype-lift

File: tests/test_output_generic.py [hunk 2]
```diff
@@ -30,7 +30,7 @@ class TestOutputGeneric(TestCase):
     @classmethod
     def setUpClass(cls) -> None:
         cls._bom = Bom()
-        cls._bom.add_component(Component(name='setuptools'))
+        cls._bom.components.add(Component(name='setuptools'))
```
Comment (jkowalleck) [inline tests/test_output_generic.py:33]: this looks so much nicer <3

File: cyclonedx/model/bom.py [hunk 3]
```diff
@@ -397,10 +302,9 @@ def has_vulnerabilities(self) -> bool:
             `bool` - `True` if at least one `cyclonedx.model.component.Component` has at least one Vulnerability,
                 `False` otherwise.
         """
-        if self.components:
-            for c in self.components:
-                if c.has_vulnerabilities():
-                    return True
+        for c in self.components:
```
Comment (jkowalleck) [inline cyclonedx/model/bom.py:305]: pythonic (untested - unittests might prove): 
```py
return any(c.has_vulnerabilities() for c in self.components)
```

**Review by jkowalleck — COMMENTED** [review]
all reviewd. adde dsome remarks.

<3 
the outcome is such a milestone.
now it is possible to use generators to create the structures and simply pass in the generators as parameters. which will make the `cyclonedx-python` likely more pythonic.

like 
```py
Bom(
components=makeComponentFromRequirement(line) for line in requirements
)
```

**Comment by jkowalleck** [comment]
also fixed #176

**Pull request merged**